### PR TITLE
Fixes/osx thick titlebar pointer events streaming, (tabs interface) #15696

### LINKED ISSUE #15696: PointerMovedEvent and PointerReleasedEvent do not trigger after mouse leaves the title bar with pressed
### Describe the bug

I add a contorl to the titlebar, and add Handlers `PointerPressedEvent`/`PointerMovedEvent`/`PointerReleasedEvent`/`PointerCaptureLostEvent`. When I press/move/release mouse left button over the control, all events trigger as expected.

If I just press mouse and move out the titlebar, `PointerMovedEvent` will not trigger, unless I release the left button. `PointerReleasedEvent` does not trigger at this point, and `e.GetCurrentPoint(this).Properties.IsLeftButtonPressed` still return `true` in `PointerMovedEvent` event, until I press left button and release next time.

But If I remove property `OSXThickTitleBar` from `ExtendClientAreaChromeHints`, everything works well.

Here is a demo video:

https://github.com/AvaloniaUI/Avalonia/assets/5343499/1efab050-820f-49d8-a93e-dd047d70e921

Here is the other demo which works perfect when I just remove `OSXThickTitleBar` property from ExtendClientAreaChromeHints:

https://github.com/AvaloniaUI/Avalonia/assets/5343499/88ca4370-eb68-4988-b843-7ea20b5a9e2b

### To Reproduce

1. Set window properties `ExtendClientAreaToDecorationsHint="True"` / `ExtendClientAreaTitleBarHeightHint="38"` / `ExtendClientAreaChromeHints="Default,OSXThickTitleBar"`
2. Add a control to the top, use the following code to listen for events
```CSharp
this.AddHandler(PointerPressedEvent, this.OnPointerPressed, RoutingStrategies.Tunnel);
this.AddHandler(PointerMovedEvent, this.OnPointerMove, RoutingStrategies.Tunnel);
this.AddHandler(PointerReleasedEvent, this.OnPointerReleased, RoutingStrategies.Tunnel);
this.AddHandler(PointerCaptureLostEvent, this.OnPointerCaptureLost, RoutingStrategies.Tunnel);
```
3. Press the control with mouse left button, and move out from titlebar, but don't leave the window.

### Expected behavior

_No response_

### Avalonia version

11.0.10

### OS

macOS

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by ErtyHackward**
## What does the pull request do?

Shortly: Allows to implement title area tabs like in Chrome with thick title bar.

This PR fixes a mouse event handling issue in Avalonia's macOS implementation when using `OSXThickTitleBar` mode. The issue prevented mouse events from being properly delivered when dragging from the titlebar area to the main window content or outside the window, breaking drag operations.

The solution implements an **event-tracking loop** pattern in `AvnWindow.mm` that intercepts mouse events starting in the titlebar area and ensures continuous event delivery to the Avalonia content view, bypassing the problematic NSToolbar event routing.

## What is the current behavior?

**Current broken behavior:**
- Mouse down events in thick titlebar area start normally
- When dragging from titlebar outside the window, mouse events stop being delivered
- AvnView never receives drag ending `mouseUp`, so drag operations fail
- Interactive controls in the main content area become unresponsive during titlebar-initiated drags
- This breaks title area controls (tabs like in Chrome)

NSToolbar in thick titlebar mode intercepts mouse events but fails to forward them when the drag moves outside its bounds, breaking the event chain that Avalonia's content view expects.

## What is the updated/expected behavior with this PR?

**Fixed behavior:**
- Mouse events starting in thick titlebar area are properly detected
- Continuous event delivery is maintained throughout drag operations from titlebar to content
- AvnView receives complete mouse event sequences (down → drag → up)
- Title are controls work seamlessly across titlebar and content boundaries
- Window dragging and all drag operations function correctly
- No impact on normal (non-titlebar) mouse event handling

**Testing:**
1. Enable `OSXThickTitleBar` in an Avalonia app
2. Place interactive control in titlebar
3. Test mouse interactions starting from titlebar area and dragging outside the window
4. Verify drag operations, button clicks, and window manipulation all work correctly

## How was the solution implemented (if it's not obvious)?

The solution uses the **event-tracking loop pattern** - a common Cocoa technique for handling mouse events that cross view boundaries:

### Key Implementation Details:

1. **Titlebar Detection (`isPointInTitlebar`):**
   - Detects when mouse down occurs in titlebar area (Y ≤ titlebarHeight)
   - Verifies the hit view is actually toolbar-related (not just coordinate-based)
   - Only activates for thick titlebar mode (`_isExtended`)

2. **Event-Tracking Loop (`sendEvent`):**

3. **Event Forwarding (`forwardToAvnView`):**
   - Forwards mouse events directly to AvnView using appropriate method calls
   - Handles all mouse button types (left, right, other)
   - Supports drag, move, up/down events

## Fixed issues

Fixes #15696

### CODE REVIEW

**Comment by cla-avalonia** [comment]
- [x] All contributors have signed the CLA.

File: native/Avalonia.Native/src/OSX/AvnWindow.mm [hunk 1]
```diff
@@ -476,8 +476,93 @@ - (NSView*) findRootView:(NSView*)view
...
+        NSEventMask mask = NSEventMaskLeftMouseDragged | NSEventMaskLeftMouseUp;
+        NSEvent *ev = event;
+        while (ev.type != NSEventTypeLeftMouseUp)
+        {
+            [self forwardToAvnView:ev];
+            ev = [NSApp nextEventMatchingMask:mask
+                                     untilDate:[NSDate distantFuture]
+                                        inMode:NSEventTrackingRunLoopMode
+                                       dequeue:YES];
+        }
```
Comment (kekekeks) [inline native/Avalonia.Native/src/OSX/AvnWindow.mm:562]: Not calling `[NSApp sendEvent:ev];` for unknown events can and **will** break other things in the framework.

Comment (kekekeks) [reply native/Avalonia.Native/src/OSX/AvnWindow.mm:562]: Also, please check if running in NSEventTrackingRunLoopMode affects our timers and animations infrastructure, IIRC there previously were issues with timers not ticking in certain run modes.

**Comment by kekekeks** [comment]
Any idea what is different about the title bar area that causes that behavior? IIRC our NSView covers it and is the top-most thing there aside from traffic lights.

**Comment by kekekeks** [comment]
Ah, NSToolbar is on the top. But IIRC we are not relying on it to provide background, so maybe we should just skip forwarding events to it? The code below the changed part is actually about forwarding events that originate from the title bar.

**Comment by ErtyHackward** [comment]
@cla-avalonia agree

**Comment by ErtyHackward** [comment]
> Ah, NSToolbar is on the top. But IIRC we are not relying on it to provide background, so maybe we should just skip forwarding events to it? The code below the changed part is actually about forwarding events that originate from the title bar.

The first solution I tried was making NSToolbar hit-test transparent, but that didn't work. NSToolbar itself isn't a view, and the internal views it creates aren't accessible through public APIs, so I couldn't override their behavior.

As a result, one of its internal views ends up capturing the mouse drag sequence and doesn't forward mouseDragged or mouseUp events — which breaks drag operations that start in the toolbar area.

If you have any idea how to cleanly make the toolbar non-interactive (or bypass this capture), I'd be happy to implement it.

File: native/Avalonia.Native/src/OSX/AvnWindow.mm (hunk 1 above)
Comment (ErtyHackward) [reply native/Avalonia.Native/src/OSX/AvnWindow.mm:562]: 1. Done
2. I've checked the following timers I found:
- Render timer (CVDisplayLink): Continues working (separate thread)
- Platform threading timer: works during the loop
- Main thread scheduling: works during the loop
- Animation system: works during the loop

I can add IntegrationTestApp modifications commit for timers if needed

**Review by drasticactions — APPROVED** [review]
As a first step, this LGTM. Stepping through the code locally, it performed as expected. 

It would be ideal if there were a direct way to prevent forwarding events to NSToolbar, but I couldn't find a reliable method to do so, either. Having the tests in place does allow for changing the implementation or adjusting for edge cases later, so I think this would be safe to implement now, and we can make adjustments later should we need to.

**Comment by huangliu** [comment]
After this update, when I tap titlebar, it will send `DoubleTapped` event.
But the double tapped event on titlebar is toggle between maximize and restore main winow, so now it changed with only single click.😂

**Comment by ErtyHackward** [comment]
It seems to be my fault. The hot fix is here btw
https://github.com/ErtyHackward/Avalonia/commit/5931215226fba363588f049f94c2f4e025f15c15
But we probably need to revert the pr and add more tests

ср, 6 авг. 2025 г. в 04:37, huangliu ***@***.***>:

**Comment by MrJul** [comment]
@ErtyHackward Could you please open a PR with your fix? This probably also needs more integration tests.

**Comment by ErtyHackward** [comment]
@MrJul opened here #19447
